File: MiniFTPServer/LibFTPServ/VFS/VFSTree.cs
using System;
using System.Linq;
using System.Collections.Generic;
using LibFTPServ.FTP;
using System.Text;

namespace LibFTPServ.VFS
{
    internal class VFSDirectory: VFSVirtualItem
    {
        private List<VFSDirectory> _subdirs;
        private List<VFSVirtualItem> _files;
        //private string _name;
        //private DateTime _date;

        private string GetFileName(string path)
        {
            string[] parts = path.Split('/');
            return parts[parts.Length - 1];
        }

        public VFSDirectory(): base(null)
        {
            _subdirs = new List<VFSDirectory>();
            _files = new List<VFSVirtualItem>();
            base.Size = -1;
        }

        public VFSDirectory(string name): base(name.ToLower())
        {
            if (string.IsNullOrEmpty(name)) throw new ArgumentException("Name can't be null or empty");
            _subdirs = new List<VFSDirectory>();
            _files = new List<VFSVirtualItem>();
            base.Size = -1;
        }

        public VFSDirectory(string name, DateTime date) : base(name.ToLower(), date, -1)
        {
            if (string.IsNullOrEmpty(name)) throw new ArgumentException("Name can't be null or empty");
            _subdirs = new List<VFSDirectory>();
            _files = new List<VFSVirtualItem>();
        }

        public bool ContainsDir(string dirname)
        {
            var q = from dir in _subdirs where dir.Name == dirname.ToLower() select dir;
            return q.ToList().Count > 0;
        }

        public VFSDirectory GetSubdirByName(string name)
        {
            var q = from dir in _subdirs where dir.Name == name.ToLower() select dir;
            return q.FirstOrDefault();
        }

        public List<FTPListItem> ListDirectory()
        {
            List<FTPListItem> ret = new List<FTPListItem>();
            FTPListItem tmp;

            foreach (VFSDirectory dir in Subdirs)
            {
                tmp = new FTPListItem();
                tmp.Name = dir.Name;
                tmp.IsDir = true;
                tmp.Date = dir.Date;
                ret.Add(tmp);
            }

            foreach (VFSVirtualItem itm in Files)
            {
                tmp = new FTPListItem();
                tmp.Name = GetFileName(itm.Name);
                tmp.Date = itm.Date;
                tmp.IsDir = false;
                tmp.Size = itm.Size;
                ret.Add(tmp);
            }

            return ret;
        }

        public List<VFSDirectory> Subdirs
        {
            get { return _subdirs; }
        }

        public List<VFSVirtualItem> Files
        {
            get { return _files; }
        }
    }
}

File: MiniFTPServer/vb/LibFTPServ.ConvertedToVBNet/VFS/VFSTree.vb
Imports System.Linq
Imports System.Collections.Generic
Imports LibFTPServ.FTP
Imports System.Text

Namespace VFS
	Friend Class VFSDirectory
		Inherits VFSVirtualItem
		Private _subdirs As List(Of VFSDirectory)
		Private _files As List(Of VFSVirtualItem)
		'private string _name;
		'private DateTime _date;

		Private Function GetFileName(path As String) As String
			Dim parts As String() = path.Split("/"C)
			Return parts(parts.Length - 1)
		End Function

		Public Sub New()
			MyBase.New(Nothing)
			_subdirs = New List(Of VFSDirectory)()
			_files = New List(Of VFSVirtualItem)()
			MyBase.Size = -1
		End Sub

		Public Sub New(name As String)
			MyBase.New(name.ToLower())
			If String.IsNullOrEmpty(name) Then
				Throw New ArgumentException("Name can't be null or empty")
			End If
			_subdirs = New List(Of VFSDirectory)()
			_files = New List(Of VFSVirtualItem)()
			MyBase.Size = -1
		End Sub

		Public Sub New(name As String, [date] As DateTime)
			MyBase.New(name.ToLower(), [date], -1)
			If String.IsNullOrEmpty(name) Then
				Throw New ArgumentException("Name can't be null or empty")
			End If
			_subdirs = New List(Of VFSDirectory)()
			_files = New List(Of VFSVirtualItem)()
		End Sub

		Public Function ContainsDir(dirname As String) As Boolean
            Dim q = From dir In _subdirs Where dir.Name = dirname.ToLower() 'dir
			Return q.ToList().Count > 0
		End Function

		Public Function GetSubdirByName(name As String) As VFSDirectory
            Dim q = From dir In _subdirs Where dir.Name = name.ToLower() 'dir
			Return q.FirstOrDefault()
		End Function

		Public Function ListDirectory() As List(Of FTPListItem)
			Dim ret As New List(Of FTPListItem)()
			Dim tmp As FTPListItem

			For Each dir As VFSDirectory In Subdirs
				tmp = New FTPListItem()
				tmp.Name = dir.Name
				tmp.IsDir = True
				tmp.[Date] = dir.[Date]
				ret.Add(tmp)
			Next

			For Each itm As VFSVirtualItem In Files
				tmp = New FTPListItem()
				tmp.Name = GetFileName(itm.Name)
				tmp.[Date] = itm.[Date]
				tmp.IsDir = False
				tmp.Size = itm.Size
				ret.Add(tmp)
			Next

			Return ret
		End Function

		Public ReadOnly Property Subdirs() As List(Of VFSDirectory)
			Get
				Return _subdirs
			End Get
		End Property

		Public ReadOnly Property Files() As List(Of VFSVirtualItem)
			Get
				Return _files
			End Get
		End Property
	End Class
End Namespace
